### PR TITLE
Correct Windows argument handling

### LINKED ISSUE #1881: Is our Windows Argument handling correct?
Windows programs receive the CLI arguments as a single long string, but Miri gets the already separated arguments from `std::env` -- so to emulate Windows programs, we have to construct the one-string version from that:

https://github.com/rust-lang/miri/blob/d4a76a9ef113f733ae992fdff080ce1bfa7c408b/src/eval.rs#L214-L223

Our tests ensure that this is at least [not entirely broken](https://github.com/rust-lang/miri/blob/master/test-cargo-miri/run-test.py#L97). However, I am not sure if this will truly handle all cases correctly.

@ChrisDenton I saw you recently fixed things here on the standard library side, would be great if you could also take a look at what Miri is doing to ensure that it is correct. :)

### PR DESCRIPTION
**Pull request by ChrisDenton**
Previously the command line string would have been incorrectly constructed if argv[0] contained a doublequote (`"`) or ended in a trailing backslash (`\`). This is a very rare edge case because, by convention, argv[0] is the path to the application and Windows file names cannot contain doublequotes.

Fixes #1881

### CODE REVIEW

File: src/eval.rs [hunk 1]
```diff
@@ -353,3 +345,76 @@ pub fn eval_entry<'tcx>(
         Err(e) => report_error(&ecx, e),
     }
 }
+
+// Turns an array of arguments into a Windows command line string.
+//
+// The string will be UTF-16 encoded and NUL terminated.
+// 
+// Panics if the zeroth argument contains the `"` character because doublequotes
+// in argv[0] cannot be encoded using the standard command line parsing rules.
```
Comment (bjorn3) [inline src/eval.rs:354]:
```suggestion
/// Turns an array of arguments into a Windows command line string.
///
/// The string will be UTF-16 encoded and NUL terminated.
/// 
/// Panics if the zeroth argument contains the `"` character because doublequotes
/// in argv[0] cannot be encoded using the standard command line parsing rules.
```

Private functions are visible in the documentation when using `--document-private-items`.

Comment (ChrisDenton) [reply src/eval.rs]: Ah thanks! Should be fixed now (along with the formatting).

**Comment by ChrisDenton** [comment]
Sorry about the noise. Had a bit of trouble with formatting properly.

**Comment by RalfJung** [comment]
Thanks a lot for the PR!

> Previously there were rare edge cases where the command line string would have been incorrectly constructed.

Can you extend this line and the reference file to check at least one of those edge cases?

https://github.com/rust-lang/miri/blob/6cf851f6d43bff1e8ac1421cbf3c6a8d20f89560/test-cargo-miri/run-test.py#L97

**Comment by ChrisDenton** [comment]
I'll update the comment but does the test allow testing the zeroth argument? That's the only one that I think has issues.

**Comment by RalfJung** [comment]
Oh... hm yeah not sure how to test that one.

File: src/eval.rs [hunk 2]
```diff
@@ -353,3 +345,76 @@ pub fn eval_entry<'tcx>(
...
+        } else {
+            // Always surround argv[0] with quotes.
+            let mut s = String::new();
+            s.push('"');
+            s.push_str(arg0);
+            s.push('"');
+            s
+        }
+    } else {
+        return vec![0];
```
Comment (RalfJung) [inline src/eval.rs:376]: Makes it a bit hard to read with the `else` branch so far from the `if let`... maybe something like this?
```rust
let arg0 = if let Some(arg0) = args.next() {
    arg0
  } else {
    // Entirely empty list of args
    return vec![0];
  };
```

**Comment by RalfJung** [comment]
> I'll update the comment but does the test allow testing the zeroth argument? That's the only one that I think has issues.

I guess the 2nd best thing we could do is have a `#[test]` inside `eval.rs` that checks `args_to_utf16_command_string` for some interesting cases.

File: src/eval.rs [hunk 3]
```diff
@@ -353,3 +345,76 @@ pub fn eval_entry<'tcx>(
...
+fn args_to_utf16_command_string<I, T>(mut args: I) -> Vec<u16>
+where
+    I: Iterator<Item = T>,
+    T: AsRef<str>,
+{
+    // Parse argv[0]. Slashes aren't escaped. Literal double quotes are not allowed.
+    let mut cmd = if let Some(arg0) = args.next() {
+        let arg0 = arg0.as_ref();
+        if arg0.is_empty() {
+            "\"\"".into()
```
Comment (RalfJung) [inline src/eval.rs:364]: Why does this need a special case, wouldn't the `else` branch below also do the right thing?

File: src/eval.rs [hunk 4]
```diff
@@ -353,3 +345,76 @@ pub fn eval_entry<'tcx>(
...
+    };
+
+    // Build the other arguments.
+    for arg in args {
+        let arg = arg.as_ref();
+        cmd.push(' ');
+        if arg.is_empty() {
+            cmd.push_str("\"\"");
+        } else if !arg.bytes().any(|c| matches!(c, b'"' | b'\t' | b' ')) {
+            cmd.push_str(arg);
```
Comment (RalfJung) [inline src/eval.rs:386]:
```suggestion
            // No quote, tab, or space -- no escaping required.
            cmd.push_str(arg);
```

File: src/eval.rs [hunk 5]
```diff
@@ -353,3 +345,76 @@ pub fn eval_entry<'tcx>(
...
+                    chars.next();
+                    nslashes += 1;
+                }
+
+                match chars.next() {
+                    Some('"') => {
+                        cmd.extend(iter::repeat('\\').take(nslashes * 2 + 1));
+                        cmd.push('"');
+                    }
+                    Some(c) => {
```
Comment (RalfJung) [inline src/eval.rs:402]: So the argument `\ ` becomes just `"\ "`, no further escaping for that backslash?

Comment (ChrisDenton) [reply src/eval.rs:402]: Yes. From [Parsing C++ command-line arguments](https://docs.microsoft.com/en-us/cpp/cpp/main-function-command-line-args?view=msvc-160#parsing-c-command-line-arguments):

> The rule is that one or more `\` only need to be escaped if they are immediately followed by `"`.

I think I should add that link as a comment as well.

Comment (RalfJung) [reply src/eval.rs:402]: Yes that is a helpful link.

From that text it is actually not clear to me that one can use `\"` inside double quote marks; the text starting "A string surrounded by double quote marks" only mentions "a pair of double quote marks" and does not say anything about `\` being special.  Also none of the examples cover this. So it is not clear to me from that text if `"a\\b"` will be interpreted as `a\\b` or `a\b` ending up in argv.

File: src/eval.rs (hunk 4 above)
Comment (RalfJung) [reply src/eval.rs:386]: The fact that we don't have to escape even when there are `\` is quite surprising and might be worth a comment; this is again that "Backslashes are interpreted literally, unless they immediately precede a double quote mark" clause.

File: src/eval.rs (hunk 5 above)
Comment (ChrisDenton) [reply src/eval.rs:402]: Perhaps this is a more informative link: https://daviddeley.com/autohotkey/parameters/parameters.htm#WINCRULES

It goes through examples and defines the rules a bit more precisely.

File: src/eval.rs (hunk 4 above)
Comment (ChrisDenton) [reply src/eval.rs:386]: Maybe I should explain this in the next else block? Something like:

```rust
} else {
    // Spaces and tabs are escaped by surrounding them in quotes.
    // Quotes are themselves escaped by using backslashes when in a
    // quoted block.
    // Backslashes only need to be escaped when one or more are directly
    // followed by a quote. Otherwise they are taken literally.
```

Comment (RalfJung) [reply src/eval.rs:386]: Sounds good!

**Comment by ChrisDenton** [comment]
Ok, I've hopefully addressed all the above issues.

The new code at the bottom is tests for the `argv[0]` edge cases.

**Comment by RalfJung** [comment]
Thanks a lot. :)
@bors r+
